store/index/index.go

```diff
@@ -90,119 +92,226 @@ const BucketPoolSize = 1024
 			return nil, err
 		}
 	} else {
 		if err != nil {
 			return nil, err
 		}
-		buckets, sizeBuckets, err = scanIndex(path, indexSizeBits)
+
+		if header.BucketsBits != indexSizeBits {
+			return nil, types.ErrIndexWrongBitSize{header.BucketsBits, indexSizeBits}
+		}
+
+		lastIndexNum, err = scanIndex(path, header.FirstFile, indexSizeBits, buckets, sizeBuckets, fileBuckets)
 		if err != nil {
 			return nil, err
 		}
-		file, err = openFileRandom(path, os.O_RDWR|os.O_APPEND|os.O_EXCL)
+
+		lastIndexPath = indexFileName(path, lastIndexNum)
+		file, err = openFileAppend(lastIndexPath)
 		if err != nil {
 			return nil, err
 		}
-		length = types.Position(stat.Size())
-	}
-	return &Index{
-		indexSizeBits,
-		buckets,
-		sizeBuckets,
-		file,
-		bufio.NewWriterSize(file, indexBufferSize),
-		primary,
-		sync.RWMutex{},
-		0,
-		make(bucketPool, BucketPoolSize),
-		make(bucketPool, BucketPoolSize),
-		length,
-	}, nil
-}
-
-func scanIndex(path string, indexSizeBits uint8) (Buckets, SizeBuckets, error) {
-	// this is a single sequential read across the whole index
-	file, err := openFileForScan(path)
-	if err != nil {
-		return nil, nil, err
 	}
-	defer func() {
-		_ = file.Close()
-	}()
-	header, bytesRead, err := ReadHeader(file)
+
+	fi, err := file.Stat()
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-	if header.BucketsBits != indexSizeBits {
-		return nil, nil, types.ErrIndexWrongBitSize{header.BucketsBits, indexSizeBits}
+
+	idx := &Index{
+		sizeBits:    indexSizeBits,
+		buckets:     buckets,
+		sizeBuckets: sizeBuckets,
+		fileBuckets: fileBuckets,
+		file:        file,
+		fileNum:     lastIndexNum,
+		headerPath:  headerPath,
+		writer:      bufio.NewWriterSize(file, indexBufferSize),
+		Primary:     primary,
+		bucketLk:    sync.RWMutex{},
+		curPool:     make(bucketPool, BucketPoolSize),
+		nextPool:    make(bucketPool, BucketPoolSize),
+		length:      types.Position(fi.Size()),
+		basePath:    path,
+		updateSig:   make(chan struct{}, 1),
+		gcDone:      make(chan struct{}),
 	}
-	buckets, err := NewBuckets(indexSizeBits)
+
+	go idx.garbageCollector()
+
+	return idx, nil
+}
+
+func indexFileName(basePath string, fileNum uint32) string {
+	return fmt.Sprintf("%s.%d", basePath, fileNum)
+}
+
+func scanIndex(basePath string, fileNum uint32, indexSizeBits uint8, buckets Buckets, sizeBuckets SizeBuckets, fileBuckets FileBuckets) (uint32, error) {
+	if len(buckets) != 1<<indexSizeBits {
+		panic("wrong size of buckets")
+	}
+	if len(sizeBuckets) != 1<<indexSizeBits {
+		panic("wrong size of sizeBuckets")
+	}
+	if len(fileBuckets) != 1<<indexSizeBits {
+		panic("wrong size of fileBuckets")
+	}
+
+	var lastFileNum uint32
+	for {
+		err := scanIndexFile(basePath, fileNum, indexSizeBits, buckets, sizeBuckets, fileBuckets)
+		if err != nil {
+			if os.IsNotExist(err) {
+				break
+			}
+			return 0, err
+		}
+		lastFileNum = fileNum
+		fileNum++
+	}
+	return lastFileNum, nil
+}
+
+func (i *Index) StorageSize() (int64, error) {
```

Comment:
Seems to only be used by `IndexStorageSize` `store.go` which itself is not used anywhere else. 

Regardless, it's certainly nice to have this; do you have future plans for using this functionality?

Reply:
It is used in the next version of go-indexer-core, because that had a hardcoded index name and needs to use this instead.